### PR TITLE
Added documentation and type-hinting to base element class. 

### PR DESCRIPTION
**Pull request by washad**

### CODE REVIEW

**Comment by washad** [comment]
Note, some of the changes are artifacts of running the "black" formatting tool before submitting my code.

**Comment by falkoschindler** [comment]
Thanks a lot for this pull request, @washad!
The diff is quite large and we'd like to fix some issues (mainly with linting) before merging.

- We use autopep8 with a maximum line width of 120 (see .vscode/settings.json). It would be great if you could setup your IDE (PyCharm, I guess) the same. And maybe we can add a similar configuration file like VSCode's settings.json to the repo, so that everybody else with PyCharm automatically uses the same settings.
- We prefer single quotes. PEP8 doesn't give a preference and we simply chose "'". (For docstrings triple double quotes are recommended. That's right in your code.)
- There's a `... note::` marker in line 249 (three dots). As far as I can tell this is not valid RST.
- I think VSCode doesn't show RST notes because they represent additional information that is not important enough for the small popup window. I'm not sure. But maybe we use them more sparingly or not at all.

You also started to improve type-hinting, which is somewhat independent of the documentation and might deserve its own PR with its own dedicated space for discussion. But here are my thoughts so far:

- We [agreed](https://github.com/zauberzeug/nicegui/discussions/350#discussioncomment-4925521) that `to_dict` deserves an underscore since it's not for public use. And within NiceGUI it's only used twice right before sending it to the client via SocketIO. Therefore I don't see the value of type-annotating its return type more then `Dict`. An `ElementAsDict` in a separate file needs to be kept in sync and is prone for regression errors.
- I'm also sceptical towards the event types. In its current form, the auto-completion does not include events from libraries like Quasar or AG Grid (e.g. "cellDoubleClicked"). And it doesn't indicate the possibility for modifiers (e.g. "keydown.shift.x"). It might make sense when individual elements define their own literal strings for the event type by overwriting the `on()` method. But then we can also leave the arbitrary string for the base implementation and add more specific events in the child classes.

Long story short: Let's concentrate on documentation and fix the formatting differences.

**Comment by washad** [comment]
I'll try to have an update this weekend : )

**Comment by washad** [comment]
I've made the requested changes, lemme know : )

**Comment by falkoschindler** [comment]
Thanks, @washad!
I reviewed it, did some minor changes here and there and merged it into main. 🙂
